### PR TITLE
[ASCII-1458] remove prometheus dependency from serverless binary using a separate telemetry implementation for serverless

### PR DESCRIPTION
**Pull request by GustavoCaso**
### What does this PR do?

We are removing the prometheus dependency from the serverless binary.

We accomplish that by creating a new `telemetry/noopsimpl` and creating types for serverless and non serverless build removing all reference Prometheus directly from the telemetry component interface. Then with the help of build flags `//go:build serverless` we can ensure the serverless binary has a noop interface for the telemetry component. 

For the record I computed some stats for the init duration of the serverless agent on this branch:

```
base:   min=478.33 ms, max=511.23 ms, avg=493.6 ms, median=490.38 ms
branch: min=459.99 ms, max=490.08 ms, avg=472.6 ms, median=471.875 ms
```

So removing some prometheus dependencies removes around 20ms of init time duration

### Describe how to test/QA your changes

Ensure telemetry data for non serverless binary gets sent correctly

### CODE REVIEW

**Review by hmahmood — APPROVED** [review]

**Review by vickenty — APPROVED** [review]

**Review by brycekahle — APPROVED** [review]

File: comp/core/telemetry/serverlessimpl/counter.go
Comment (pgimalac) [inline comp/core/telemetry/serverlessimpl/counter.go:1]: Do we want to call it `serverlessimpl`, or more something like `noopimpl` ?
(I just feel like it doesn't have anything serverless-specific)

File: comp/core/telemetry/serverlessimpl/counter.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build serverless
```
Comment (pgimalac) [inline comp/core/telemetry/serverlessimpl/counter.go:6]: I guess we can remove the `serverless` tags ?

File: comp/core/telemetry/serverlessimpl/telemetry.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+var dummyHandler = dummy{}
+
+func (t *serverlessImpl) Handler() http.Handler {
+	return &dummyHandler
+}
+
+func (t *serverlessImpl) Reset() {
+	// NADA
+}
```
Comment (pgimalac) [inline comp/core/telemetry/serverlessimpl/telemetry.go:39]:
```suggestion
func (t *serverlessImpl) Reset() {}
```

File: comp/core/telemetry/telemetryimpl/prom_counter.go [hunk 3]
```diff
@@ -3,6 +3,8 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016-present Datadog, Inc.
 
+//go:build !serverless
```
Comment (pgimalac) [inline comp/core/telemetry/telemetryimpl/prom_counter.go:6]: I guess we also don't need the `!serverless` tags here ?

File: pkg/network/tracer/tracer.go [hunk 4]
```diff
@@ -161,8 +161,7 @@ func newTracer(cfg *config.Config) (_ *Tracer, reterr error) {
 	}()
 
 	if tr.bpfErrorsCollector = ebpftelemetry.NewEBPFErrorsCollector(); tr.bpfErrorsCollector != nil {
-		coretelemetry.GetCompatComponent().RegisterCollector(tr.bpfErrorsCollector)
-
+		coretelemetry.GetCompatComponent().(coretelemetry.PrometheusComponent).RegisterCollector(tr.bpfErrorsCollector)
```
Comment (pgimalac) [inline pkg/network/tracer/tracer.go:164]: I feel like having to cast the interface to a specific impl goes against the "component way" of doing things...
And it also means that the implementations are not interchangeable

File: comp/core/telemetry/telemetryimpl/prom_counter.go [hunk 5]
```diff
@@ -3,8 +3,6 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016-present Datadog, Inc.
 
-//go:build !serverless
```
Comment (truthbk) [inline comp/core/telemetry/telemetryimpl/prom_counter.go]: I presume you were able to validate the compiler is able to figure this out and there's no trace of `prometheus` in the build despite not having a build tag in these parts of the code?

**Review by truthbk — APPROVED** [review]
👍

**Review by ahmed-mez — COMMENTED** [review]
The changes in the trace-agent are exclusively about the OTLP receiver - can we request a review from @DataDog/opentelemetry ?

File: comp/trace/config/otlp_serverless.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+)
+
+func setOtlpReceiver(c *config.AgentConfig, core corecompcfg.Component, grpcPort int) error {
+	c.OTLPReceiver = &config.OTLP{
+		BindHost:               c.ReceiverHost,
+		GRPCPort:               grpcPort,
+		MaxRequestBytes:        c.MaxRequestBytes,
+		SpanNameRemappings:     coreconfig.Datadog.GetStringMapString("otlp_config.traces.span_name_remappings"),
+		SpanNameAsResourceName: core.GetBool("otlp_config.traces.span_name_as_resource_name"),
+		ProbabilisticSampling:  core.GetFloat64("otlp_config.traces.probabilistic_sampler.sampling_percentage"),
```
Comment (dineshg13) [inline comp/trace/config/otlp_serverless.go:23]: Why are we not setting attributesTranslator ?

Comment (GustavoCaso) [reply comp/trace/config/otlp_serverless.go:23]: I believe I was wrong on doing this change on this PR. This PR focused on removing prometheus from the serverless binary. In this case these changes are more related to otlp which should be handle on a separate PR

Comment (GustavoCaso) [reply comp/trace/config/otlp_serverless.go]: Done https://github.com/DataDog/datadog-agent/pull/24854/commits/e6cf9f78c89920af3125cbcf051011dd659f4739

**Review by dineshg13 — APPROVED** [review]

**Comment by GustavoCaso** [comment]
For the record I computed some stats for the init duration of the serverless agent on this branch:

```
base:   min=478.33 ms, max=511.23 ms, avg=493.6 ms, median=490.38 ms
branch: min=459.99 ms, max=490.08 ms, avg=472.6 ms, median=471.875 ms
```
So removing some prometheus dependencies removes around 20ms of init time duration

**Review by ahmed-mez — APPROVED** [review]

**Pull request merged**